### PR TITLE
Synchronize files from linux-system-roles/template

### PR DESCRIPTION
**Pull request by systemroller**
This PR propagates files from [linux-system-roles/template](https://github.com/linux-system-roles/template) which should be in sync across [linux-system-roles](https://github.com/linux-system-roles) repos. In case of changing affected files via pushing to this PR, please do not forget also to push the changes to [linux-system-roles/template](https://github.com/linux-system-roles/template) repo.

CC: @i386x, @pcahyna.

### CODE REVIEW

**Comment by pcahyna** [comment]
@i386x "Files copied on wrong place " - is this a bug in the script or in the template?

**Comment by tyll** [comment]
The custom command fails on Travis: https://travis-ci.com/linux-system-roles/network/jobs/258652534#L272

File: tox.ini [hunk 1]
```diff
@@ -247,9 +260,10 @@ max-line-length = 88
...
-  3.6: py36,black,extra
-  3.7: py37,extra
-  3.8: py38,extra
+  2.6: py26,custom
+  2.7: py27,flake8,pylint,custom
+  3.5: coveralls,custom
+  3.6: py36,black,coveralls,custom
+  3.7: py37,coveralls,custom
+  3.7-dev: py37,coveralls,custom
+  3.8-dev: py38,coveralls,custom
```
Comment (tyll) [inline tox.ini]: Is it correct to remove python 3.8 here?

**Comment by pcahyna** [comment]
> The custom command fails on Travis: https://travis-ci.com/linux-system-roles/network/jobs/258652534#L272

`.travis/custom.sh: line 26: .tox/env-3.6/bin/python: No such file or directory`
maybe it needs python3 instead of python?

**Comment by i386x** [comment]
> @i386x "Files copied on wrong place " - is this a bug in the script or in the template?

This was a bug in sync script. It should be fixed now.

**Comment by i386x** [comment]
> > The custom command fails on Travis: https://travis-ci.com/linux-system-roles/network/jobs/258652534#L272
> 
> `.travis/custom.sh: line 26: .tox/env-3.6/bin/python: No such file or directory`
> maybe it needs python3 instead of python?

This puzzles me most, the absolute path `/home/travis/build/linux-system-roles/network/.tox/env-3.6/bin/python`, passed to `custom.sh` script as 1st argument, was shortened to `.tox/env-3.6/bin/python` in some way, which results in the mentioned error, because `.tox/env-3.6/bin/python` cannot be seen from `tests/` directory.

File: .travis/custom.sh [hunk 2]
```diff
@@ -23,5 +23,5 @@ shift 2
 
 # Write your custom commands here that should be run when `tox -e custom`:
 if [[ -z "${TRAVIS}" ]] || lsr_check_python_version ${ENVPYTHON} -eq '3.6'; then
-  (set -x; cd ${TOPDIR}/tests && ${ENVPYTHON} ./ensure_non_running_provider.py)
+  (set -x; cd ${TOPDIR}/tests; ${ENVPYTHON} ./ensure_non_running_provider.py)
```
Comment (pcahyna) [inline .travis/custom.sh]: why this change? you don't care whether cd has succeeded or failed?

File: .travis/runcoveralls.sh [hunk 3]
```diff
@@ -56,8 +56,8 @@ source =
     .
     ${TOPDIR}
 EOF
-mv ${TOPDIR}/.coverage ${TOPDIR}/.coverage.merge
-${ENVPYTHON} -m coverage combine --append ${TOPDIR}
+mv ${TOPDIR}/.coverage ${TOPDIR}/.coverage.merge || :
+${ENVPYTHON} -m coverage combine --append ${TOPDIR} || :
```
Comment (pcahyna) [inline .travis/runcoveralls.sh]: @tyll can you please review the Coveralls bits?

File: .travis/custom.sh (hunk 2 above)
Comment (i386x) [reply .travis/custom.sh]: `set -e` should handle this (exit immediately if a command exits with a non-zero status).

**Comment by i386x** [comment]
> > > The custom command fails on Travis: https://travis-ci.com/linux-system-roles/network/jobs/258652534#L272
> > 
> > 
> > `.travis/custom.sh: line 26: .tox/env-3.6/bin/python: No such file or directory`
> > maybe it needs python3 instead of python?
> 
> This puzzles me most, the absolute path `/home/travis/build/linux-system-roles/network/.tox/env-3.6/bin/python`, passed to `custom.sh` script as 1st argument, was shortened to `.tox/env-3.6/bin/python` in some way, which results in the mentioned error, because `.tox/env-3.6/bin/python` cannot be seen from `tests/` directory.

Not bash, but https://github.com/tox-dev/tox/commit/43d65b17#diff-57c0f90ac19d5a79d6495ab22fab6539R247-R258 is the source of problem (luckily, workaround should be easy).

**Comment by pcahyna** [comment]
> > > > The custom command fails on Travis: https://travis-ci.com/linux-system-roles/network/jobs/258652534#L272
> > > 
> > > 
> > > `.travis/custom.sh: line 26: .tox/env-3.6/bin/python: No such file or directory`
> > > maybe it needs python3 instead of python?
> > 
> > 
> > This puzzles me most, the absolute path `/home/travis/build/linux-system-roles/network/.tox/env-3.6/bin/python`, passed to `custom.sh` script as 1st argument, was shortened to `.tox/env-3.6/bin/python` in some way, which results in the mentioned error, because `.tox/env-3.6/bin/python` cannot be seen from `tests/` directory.
> 
> Not bash, but [tox-dev/tox@43d65b1#diff-57c0f90ac19d5a79d6495ab22fab6539R247-R258](https://github.com/tox-dev/tox/commit/43d65b17#diff-57c0f90ac19d5a79d6495ab22fab6539R247-R258) is the source of problem (luckily, workaround should be easy).

Is this behavior of tox documented? Should it be reported as a bug if not?

**Comment by i386x** [comment]
> Is this behavior of tox documented? Should it be reported as a bug if not?

I file issue on [tox-dev/tox](https://github.com/tox-dev/tox) for it: https://github.com/tox-dev/tox/issues/1463

File: tox.ini (hunk 1 above)
Comment (i386x) [reply tox.ini]: I am not sure. Python 3.8.0 is the latest release of Python now and will be shortly supported in [Fedora](https://fedoraproject.org/wiki/Changes/Python3.8). Also it seems that it is [supported](https://docs.travis-ci.com/user/languages/python/) by Travis, so I propose to add `3.8` line besides `3.8-dev` line.

**Comment by pcahyna** [comment]
> > Is this behavior of tox documented? Should it be reported as a bug if not?
> 
> I file issue on [tox-dev/tox](https://github.com/tox-dev/tox) for it: [tox-dev/tox#1463](https://github.com/tox-dev/tox/issues/1463)

Why is conversion of absolute to relative path even a problem? The result should be equivalent, no?

**Comment by i386x** [comment]
> Why is conversion of absolute to relative path even a problem? The result should be equivalent, no?

The problem is a little bit complex. Seeing the logs, sometimes the path to virtual env interpreter is [absolute](https://travis-ci.com/linux-system-roles/network/jobs/258652535#L223) and sometimes it is relative. Moreover, the path shortening feature can be changed or reverted in the future so it is good to not rely on it.

File: .travis/runblack.sh [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  exit 0
+fi
+
+# Sanitize path in case if running within tox (see
+# https://github.com/tox-dev/tox/issues/1463):
+ENVPYTHON=$(readlink -f $1)
+shift
+
+DEFAULT_INCLUDE='^[^.].*\.py$'
+DEFAULT_EXCLUDE='/(\.[^.].*|tests/roles)/'
```
Comment (tyll) [inline .travis/runblack.sh]: I would prefer to run black for the python scripts in the tests dir as well.

File: .travis.yml [hunk 5]
```diff
@@ -1,7 +1,10 @@
 # SPDX-License-Identifier: MIT
 ---
-dist: xenial
```
Comment (tyll) [inline .travis.yml]: Why is this removed? Xenial seems to be the current default, so it does not change anything now but will once Travis changes the default. We could move to bionic to get Python 3.6 as the system Python removing the need for Python 3.5.

File: tox.ini (hunk 1 above)
Comment (pcahyna) [reply tox.ini]: do we need both 3.7-dev and 3.8-dev ? also, do we need coveralls in multiple places?

File: .lgtm.yml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+extraction:
+  python:
+    python_setup:
+      version: 2
```
Comment (tyll) [inline .lgtm.yml]: I merged this file into master now.

**Comment by tyll** [comment]
I merged some changes in #136 to demonstrate what I would wish the commits look like.

File: .travis/runtox [hunk 7]
```diff
@@ -1,16 +1,38 @@
...
+
+if lsr_check_python_version python -le '3.4'; then
+  lsr_info "${ME}: Only Python >= 3.5 is allowed to install Ansible. Skipping."
+  exit $E
 fi
 
-tox "$@"
-for X in ${LSR_MSCENARIOS}; do
-  LSR_MSCENARIO=${X} tox -e molecule
+for X in ${LSR_ANSIBLES}; do
```
Comment (tyll) [inline .travis/runtox]: IMHO it would be better to avoid variables like `X` and `Y` and use for example `ansible_dependency` and `scenario`.

File: .travis/preinstall [hunk 8]
```diff
@@ -1,18 +1,32 @@
...
+set -e
 
-if [[ -f ${CONFIG} ]]; then
-  . ${CONFIG}
+SCRIPTDIR=$(readlink -f $(dirname $0))
+
+# Include library.
+. ${SCRIPTDIR}/utils.sh
+
+# Add python3-selinux package (it turned out that in some circumstances, i.e
```
Comment (tyll) [inline .travis/preinstall]: AFAIU this is because molecule is using the copy or file Ansible modules on a selinux enabled system to setup the container. It has nothing to do with the tests itself but with setting up the test environment.

File: .travis/preinstall [hunk 9]
```diff
@@ -1,18 +1,32 @@
...
+# if selinux policy is enforcing, selinux is required by molecule even if
+# running simple test).
+if lsr_compare_pythons python -eq /usr/bin/python3; then
+  LSR_EXTRA_PACKAGES='python3-selinux'
 fi
 
+# Include config.
```
Comment (tyll) [inline .travis/preinstall]: Actually the comments about including config or utils could be removed because the actual code to do this is simple enough that the comments do not explain something new.

File: .travis/runblack.sh (hunk 4 above)
Comment (i386x) [reply .travis/runblack.sh]: Python scripts in `tests/` should be analyzed by black. The `tests/roles` part of regexp excludes directory containing symlinks.

File: .travis.yml (hunk 5 above)
Comment (i386x) [reply .travis.yml]: It was removed because one reviewer noted that this is default, so it needn't to be noted explicitly. But if intention of explicitness here was to prevent a surprise I will add it back.

File: .travis/preinstall (hunk 8 above)
Comment (i386x) [reply .travis/preinstall]: Thank You for noting this, I will mention it in the comment.

File: .travis.yml (hunk 5 above)
Comment (tyll) [reply .travis.yml]: IMHO it is better to keep it in there to make the choice explicit.

File: .travis/runtox [hunk 10]
```diff
@@ -21,18 +21,25 @@ BANNERSIZE=90
 . ${SCRIPTDIR}/config.sh
 
 lsr_banner "tox" ${BANNERSIZE}
-(set -x; tox); E=$?
+(set -x; tox); error_code=$?
```
Comment (tyll) [inline .travis/runtox]: Great that you fixed this one, too.

File: .travis.yml [hunk 11]
```diff
@@ -2,6 +2,10 @@
 ---
 dist: xenial
 language: python
+env:
+  global:
+    - LSR_ANSIBLES='ansible==2.6.* ansible==2.7.* ansible==2.8.*'
```
Comment (tyll) [inline .travis.yml]: Why is this configured in travis.yml and not in config.sh? In case somone wants to run molecule outside of travis, this needs to be defined as well, doesn't it? This also affects the variable on the next line.

File: .travis/runblack.sh [hunk 12]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+# SPDX-License-Identifier: MIT
+
+# Run black (the Python formatter). The first script argument is a path to
```
Comment (tyll) [inline .travis/runblack.sh]: Can you please add the reason why we need this wrapper to the comment?

File: .travis/runflake8.sh [hunk 13]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+# SPDX-License-Identifier: MIT
+
+# Run flake8. The first script argument is a path to Python interpreter, the
```
Comment (tyll) [inline .travis/runflake8.sh]: Can you please add the reason why we need this wrapper to the comment?

File: .travis/runpytest.sh [hunk 14]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+# SPDX-License-Identifier: MIT
+
+# Wrapper around pytest. First argument is a path to environment python, the
```
Comment (tyll) [inline .travis/runpytest.sh]: Can you please add the reason why we need this wrapper to the comment?

File: tox.ini (hunk 1 above)
Comment (tyll) [reply tox.ini]: merged the python version change in #137

File: tox.ini [hunk 15]
```diff
@@ -229,7 +262,8 @@ max-line-length = 88
 python =
   2.6: py26,custom
   2.7: py27,flake8,pylint,custom
-  3.5: coveralls,molecule,custom
```
Comment (tyll) [inline tox.ini]: It seems that molecule is removed here which is wrong, isn't it?

File: .travis/runpytest.sh (hunk 14 above)
Comment (pcahyna) [reply .travis/runpytest.sh]: @i386x

File: .travis.yml (hunk 11 above)
Comment (i386x) [reply .travis.yml]: `config.sh` should contain only parts that may vary across repositories. Running molecule outside Travis can be performed via `LSR_ANSIBLE_DEP=<ansible-dep> LSR_MSCENARIO=<molecule scenario> tox -e molecule` (just `tox -e molecule` runs molecule with the newest Ansible Python package installed and with the default scenario). Furthermore, running molecule for all supported Ansible versions and for all supported scenarios outside of Travis by one command will be imho possible only by introducing a shell script that wraps `tox` command (this possibility is worth of discussion). Maybe it should be good to write a documentation for our Travis and `tox` based CI and put it to file named like `TOX.INI.README.md`.

File: .travis/runpytest.sh (hunk 14 above)
Comment (i386x) [reply .travis/runpytest.sh]: Done in #144.

File: .travis/runblack.sh (hunk 12 above)
Comment (i386x) [reply .travis/runblack.sh]: Done in #145.

File: .travis/runflake8.sh (hunk 13 above)
Comment (i386x) [reply .travis/runflake8.sh]: Done in #147.

**Comment by i386x** [comment]
Closing this issue as it has been resolved by #142, #144, #145, #146, #147, #148, #149, and #151.
